src/firmware/app/control/trajectory_planner.c

```diff
@@ -0,0 +1,254 @@
+        // further computing)
+        double t = shared_polynomial2d_getTValueAtArcLengthOrder3(
+            path, i * arc_segment_length, arc_length_param);
+
+        // Get the X and Y position at the 't' value defined by the arc length
+        traj_elements[i].position = shared_polynomial2d_getValueOrder3(path, t);
+
+
+        // Create the polynomial representing path curvature
+        //                                              1
+        //                              ---------------------------------
+        //                                     abs(x'y'' - y'x'')
+        //        radius of curvature =      ----------------------
+        //                                     (x'^2 + y'^2)^(3/2)
+        //
+        const double numerator =
+            fabs(shared_polynomial1d_getValueOrder2(first_deriv.x, t) *
+                     shared_polynomial1d_getValueOrder1(second_deriv.y, t) -
+                 shared_polynomial1d_getValueOrder2(first_deriv.y, t) *
+                     shared_polynomial1d_getValueOrder1(second_deriv.x, t));
+        const double denominator =
+            pow(pow(shared_polynomial1d_getValueOrder2(first_deriv.x, t), 2) +
+                    pow(shared_polynomial1d_getValueOrder2(first_deriv.y, t), 2),
+                3.0 / 2.0);
+
+        const double radius_of_curvature = 1 / (numerator / denominator);
+
+        max_allowable_speed_profile[i] =
+            sqrt(max_allowable_acceleration * radius_of_curvature);
+    }
+
+    // First point on the velocity profile is always the current speed
+    velocity_profile[0] = initial_speed;
+
+    // Loop through the path forwards to ensure that the maximum velocity limit defined by
+    // the curvature is not breached by constant max acceleration of the robot (if it is,
+    // pull down the acceleration)
+    for (unsigned int i = 1; i < num_segments; i++)
+    {
+        // Vf = sqrt( Vi^2 + 2*constant_segment_length*max_acceleration)
+        const double temp_vel = sqrt(pow(velocity_profile[i - 1], 2) +
+                                     2 * arc_segment_length * max_allowable_acceleration);
+
+        // If the new speed is greater than the max allowable, reduce it to the max
+        if (temp_vel > max_allowable_speed_profile[i])
+        {
+            velocity_profile[i] = max_allowable_speed_profile[i];
+        }
+
+        else
+        {
+            // If the new speed is greater than the maximum limit of the robot, reduce it
+            // to the limit
+            if (temp_vel >= max_allowable_speed)
+            {
+                velocity_profile[i] = max_allowable_speed;
+            }
+            else
+            {
+                velocity_profile[i] = temp_vel;
+            }
+        }
+    }
+
+    // Now check backwards continuity. This is done to guarantee the robot can deccelerate
+    // in time to not breach the maximum allowable speed defined by the path curvature
```

Comment:
In general, how do we deal with cases where we are passed an infeasible path? That seems like a fairly crucial case.

Reply:
The plan was to have the high-level path planner check for infeasible cases. In the case that an infeasible path is passed to this trajectory planner the robot will be given a curve to follow (which it will to the best of it's ability). It won't fail or anything

Reply:
How would the high level path planner check for infeasible cases if we only know if we violate backwards continuity here?

Reply:
Perhaps let's have this return an enum like:
```
enum TrajectoryGenerationErrorCode {
OK,
ERR_VIOLATED_CONSTRAINTS,
};
```

Reply:
Added enum for generation status